### PR TITLE
Test activejob metrics

### PR DESCRIPTION
**Pull request by cat5inthecradle**
## Warning!!

The [AP CSP Create Performance Task](https://apcentral.collegeboard.org/courses/ap-computer-science-principles/exam) is in progress. The most critical dates are from April 3 - April 30, 2024. Please consider any risk introduced by this PR that could affect our students taking AP CSP. Code.org students taking AP CSP primarily use App Lab for their Create Task, however a small percent use Game Lab. Carefully consider whether your change has any risk of alterering, changing, or breaking anything in these two labs. Even small changes, such as a different button color, are considered significant during this time period. Reach out to the Student Learning team or Curriculum team for more details.

---

Fix bugs (replacing #57845) and add unit tests around the metrics logging callbacks in ApplicationJob.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/test/testing/includes_metrics.rb [hunk 1]
```diff
@@ -49,7 +49,7 @@ def initialize(metric_name, dimensions)
       def matches?(available_parameters)
         actual_metrics = available_parameters.shift
         @expected_dimensions.all? do |expected_dimension_name, expected_value|
-          sought_metric = actual_metrics.find {|m| m[:metric_name] == @expected_metric_name}
+          sought_metric = actual_metrics.find {|m| m[:metric_name].to_sym == @expected_metric_name}
```
Comment (cat5inthecradle) [inline dashboard/test/testing/includes_metrics.rb:52]: This and the other addition of `to_sym` were necessary to support Metrics with String names.

File: dashboard/app/jobs/application_job.rb [hunk 2]
```diff
@@ -44,10 +44,9 @@ def self.common_dimensions
         value: job_count,
         unit: 'Count',
         timestamp: Time.now,
-        dimensions: {
-          name: 'Environment',
-          value: CDO.rack_env
-        },
+        dimensions: [
+          {name: 'Environment', value: CDO.rack_env},
+        ]
```
Comment (cat5inthecradle) [inline dashboard/app/jobs/application_job.rb:49]: This actually supports arrays or single hashes, but our `includes_metrics` test tooling only supports array `dimensions`, so I'm taking the easy way out by wrapping this.

File: dashboard/test/jobs/application_job_test.rb [hunk 3]
```diff
@@ -0,0 +1,144 @@
+require 'test_helper'
+require 'testing/includes_metrics'
+
+# Testing Docs for ActiveJob 6: https://api.rubyonrails.org/v6.0.6.1/classes/ActiveJob/TestHelper.html#method-i-perform_enqueued_jobs
+class ApplicationJobTest < ActiveJob::TestCase
+  # ApplicationJob is a base class for all jobs in the application.
+  # Create a testable job class that inherits from ApplicationJob.
+  class TestableJob < ApplicationJob
+    # Using the test adapter to simulate enqueuing jobs (vs `async` or `inline`)
+    self.queue_adapter = :test
```
Comment (davidsbailey) [inline dashboard/test/jobs/application_job_test.rb]: optional: remove these lines? local experimentation suggests that this is already the default within unit tests when using ActiveJob::TestCase.

**Review by davidsbailey — APPROVED** [review]
looks great, thanks for the thorough coverage!

File: dashboard/test/jobs/application_job_test.rb (hunk 3 above)
Comment (cat5inthecradle) [reply dashboard/test/jobs/application_job_test.rb]:
```suggestion
```

I'll give it a try and see if it gets through drone.

**Comment by davidsbailey** [comment]
@cat5inthecradle heads up that this (or a previous PR) is now introducing an error in honeybadger in the test environment: https://app.honeybadger.io/projects/3240/faults/106811856

while this PR was merged around the time the errors started, so was this other PR which enabled HB errors to be captured in the test environment: https://github.com/code-dot-org/code-dot-org/pull/57662

**Comment by davidsbailey** [comment]
to clarify, that PR attempts to enable error logging only during DTTs, so my bet would be that this indicates a problem during UI tests

**Pull request merged**